pkg/controllers/routereflector/topologies/multi.go

```diff
@@ -0,0 +1,318 @@
+		rrConfig.Spec = apiv3.BGPPeerSpec{
+			NodeSelector: selector,
+			PeerSelector: selector,
+		}
+
+		toRefresh = append(toRefresh, rrConfig)
+	} else {
+		toKeep[rrConfig.GetName()] = true
+	}
+
+	// Used for round robin RR selection
+	rrIndex := -1
+	rrIndexPerZone := map[string]int{}
+	zones := []string{}
+	rrPerZone := map[string][]*corev1.Node{}
+
+	// Create per zone RR lists for MZR selection
+	if t.Config.ZoneLabel != "" {
+		for i, rr := range routeReflectors {
+			rrZone := rr.GetLabels()[t.Config.ZoneLabel]
+			log.Debugf("RR:%s's zone: %s", rr.GetName(), rrZone)
+			rrPerZone[rrZone] = append(rrPerZone[rrZone], routeReflectors[i])
+		}
+
+		for zone := range rrPerZone {
+			zones = append(zones, zone)
+		}
+		sort.Strings(zones)
+	}
+
+	// Let's find RRs for each nodes
+	for _, n := range nodeList {
+		if t.IsRouteReflector(string(n.GetUID()), n.GetLabels()) {
+			continue
+		}
+
+		routeReflectorsForNode := []*corev1.Node{}
+
+		// Run MZR selection first for nodes if zone label is configured
+		if t.Config.ZoneLabel != "" {
+			nodeZone := n.GetLabels()[t.Config.ZoneLabel]
+			log.Debugf("Node's zone: %s", nodeZone)
+
+			// Select the 1st RR from the same zone if there're any
+			if len(rrPerZone[nodeZone]) > 0 {
+				rr := selectRRfromZone(rrIndexPerZone, rrPerZone, nodeZone)
+				log.Debugf("Adding %s as 1st RR for Node:%s", rr.GetName(), n.GetName())
+				routeReflectorsForNode = append(routeReflectorsForNode, rr)
+			}
+
+			// Select the 2nd RR from a different zone
+			for _, zone := range zones {
+				if zone != nodeZone {
+					rr := selectRRfromZone(rrIndexPerZone, rrPerZone, zone)
+					log.Debugf("Adding %s as 2nd RR for Node:%s", rr.GetName(), n.GetName())
+					routeReflectorsForNode = append(routeReflectorsForNode, rr)
+					break
+				}
+			}
+		}
+
+		// Calcualte missing RR peers for node
+		peers := int(math.Min(float64(len(routeReflectors)), float64(t.Config.ReflectorsPerNode)))
+
+		// Selecting the remaning RRs sequentially with round robin
+		for len(routeReflectorsForNode) < peers {
+			rrIndex++
+			if rrIndex == len(routeReflectors) {
+				// Jump back to first
+				rrIndex = 0
+			}
+
+			rr := routeReflectors[rrIndex]
+			if isAlreadySelected(routeReflectorsForNode, rr) {
+				continue
+			}
+
+			log.Debugf("Adding %s to RRs of %s", rr.GetName(), n.GetName())
+			routeReflectorsForNode = append(routeReflectorsForNode, rr)
+		}
+
+		// Node selector template
+		nodeSelector := fmt.Sprintf("%s=='%s'", t.Config.HostnameLabel, n.GetLabels()[t.Config.HostnameLabel])
+
+		// Generate or update BGP peer config for node
+		for _, rr := range routeReflectorsForNode {
+			rrID := getRouteReflectorID(string(rr.GetUID()))
+			name := fmt.Sprintf(DefaultRouteReflectorClientName+"-%s", rrID, n.GetUID())
+
+			clientConfig := findBGPPeer(existingPeers, name)
+			if clientConfig != nil {
+				// The rule is already exists, content check not necessary because they are unique
```

Comment:
I think a content check _is_ appropriate here, since it's possible the BGPPeer has been modified or is out of date.

We should try to write failsafes into the code either way - unexpected scenarios tend to show up all over the place in controllers like this, so having a backup check which confirms the BGP peer has the right contents is a good idea. 

A good example is if, in the future, we ever have to modify the BGPpeer template used by this controller, then upgrading the controller needs to trigger an update of all the BGPPeers that it has already programmed.